### ISSUE TITLE
Please make Kubernetes example working

### ISSUE THREAD

**Issue by epicsoft** [zalando/patroni#574]
https://github.com/zalando/patroni/tree/master/kubernetes
Patroni in Kubernetes looks amaaazing. You guys rock!!!

Is it possible to adopt it to real-life conditions? It works with emptyDir: {} only.
I spend some hours trying to make it work with Kubernetes persistent volumes - no success.
	a) It does not allow to mount any pv as it is. The error is that pv is mounted as root and patroni requires postgress user.
	b) It starts and syncs fine if I add "securityContext: fsGroup: 999". But fails after first first POD delete. Postgres expects 0700 for its /data directory (which is modified by the fsGroup: 999). And it complains that postgres user password is different from patroni expects…

Would you please advise - how may I fix the issue?
I rolled back to a single Postgres instance for now :(

**Comment by CyberDem0n** [comment]
> It does not allow to mount any pv as it is. The error is that pv is mounted as root and patroni requires postgress user.

For our self we solved it very simple way. We start container as a root and change of data_dir content before starting Patroni: https://github.com/zalando/spilo/blob/feature/minispilo-10/postgres-appliance/launch.sh#L24

> And it complains that postgres user password is different from patroni expects…

This is very strange, passwords should come from Kubernetes Secrets.

> Would you please advise - how may I fix the issue?

You may have a look how we do it in [Spilo](https://github.com/zalando/spilo/tree/feature/minispilo-10/postgres-appliance) and also have a look on [postgres-operator](https://github.com/zalando-incubator/postgres-operator)

Where do you run Kubernetes? AFAIK there are some unsolved issues on OpenShift: https://github.com/zalando/patroni/issues/572 https://github.com/zalando/patroni/pull/500#issuecomment-349970894

**Comment by CyberDem0n** [comment]
> Your Kubernetes example looks very small and simple - exactly what we need. No mess of containers compared to stolon or crunchy solutions. Spilo looks a little heavy as well.

This example actually serves the only purpose - try it on minikube. The real world is is far more complicated. That's why Patroni is not dealing with filesystem permissions. Spilo actually is not that heavy, it contains a few postgres versions + some useful extensions Patroni + wal-e. Off course you can always fork it and strip out the stuff you don't need.

> Is there a possibility you update your Kubernetes example to make it work with PV? There was an attempt already - I see you commented out the section with volumeClaimTemplates :)))

This manifest came from Spilo: https://github.com/zalando/spilo/blob/master/kubernetes/spilo_kubernetes_aws.yaml. So it wasn't and attempt, but working solution. The trick is to change permissions to postgres:postgres before starting up Patroni.

**Comment by epicsoft** [comment]
Thank you! Do you have any good tutorial how to install Silo on Kubernetes? Does it require etcd to be installed?
https://github.com/zalando/patroni/tree/master/kubernetes example works fine w/o any prerequisites - just cubectl create -f  and it works!

**Comment by CyberDem0n** [comment]
I think you are looking for a Patroni helm-chart: https://github.com/kubernetes/charts/tree/master/incubator/patroni. The one from incubator requires etcd, but there is another one, which supports kubernetes-native patroni: https://github.com/unguiculus/charts/tree/feature/patroni/incubator/patroni

BTW, do you aware about [SLA](https://cloud.google.com/kubernetes-engine/sla)? It is only 99.5%.
When Kubernetes API is not accessible, Patroni will restart postgres in read-only...

**Comment by unguiculus** [comment]
Thanks for pointing to my chart update. I did not pursue it any further and am not sure in what state I left it. You may try it out, though. I wanted to try out the patroni-operator next but just haven't had the time yet.

**Comment by epicsoft** [comment]
> BTW, do you aware about SLA? It is only 99.5%.

Does it correspond to  kubernetes-native or both solutions?

**Comment by epicsoft** [comment]
Would you please tell how to enable TLS in Patroni helm-chart?

**Comment by CyberDem0n** [comment]
> Does it correspond to kubernetes-native or both solutions?

This is SLA of Kubernetes API from GCE. Unfortunately Patroni can't give you more.
In case if Kubernetes API is not available, Patroni will restart master in read-only mode.

> Would you please tell how to enable TLS in Patroni helm-chart?

What kind of TLS to you mean? Do you want to enable ssl in postgres?
By default Spilo generates self-signed certificates, but it's possible to provide your own certificate and private key in environment variables: 'SSL_CERTIFICATE', 'SSL_PRIVATE_KEY'.
I am not sure that helm-charts supports that, but it's shouldn't be that hard to fix it.

**Comment by epicsoft** [comment]
Yes. I'm talking about SSL for Postgres endpoint (port 5432). So, by default it is SSL protected with self-signed certificate, right?

I found nothing regarding SSL in the documentation here: http://patroni.readthedocs.io
So, I need to just set SSL_CERTIFICATE and SSL_PRIVATE_KEY ENV for the unguiculus/patroni image?
Is it path to the ssl files oк actual value of certificate/key?

**Comment by CyberDem0n** [comment]
>  So, by default it is SSL protected with self-signed certificate, right?

Correct.

> I found nothing regarding SSL in the documentation here: http://patroni.readthedocs.io

Thats true, because Patroni is not solving this problem for you. But it has all control on postgres configuration. Basically it is your task to provide certificates and configure and postgres to use them.

> So, I need to just set SSL_CERTIFICATE and SSL_PRIVATE_KEY ENV for the unguiculus/patroni image?
Is it path to the ssl files or actual values of certificate/key?

You have to put a content of certificate and private key into environment. See implementation as a reference: https://github.com/zalando/spilo/blob/master/postgres-appliance/configure_spilo.py#L56-L59

**Comment by epicsoft** [comment]
Good. Don't want be very annoying, but would you please confirm that setting these ENVs is all I need. I may see there is a check elif section == 'certificate': (it is probably from args parsing). So, I want to be sure this condition is executed for the unguiculus/patroni image (or I need to add some args into the statefulset.yaml).

**Comment by CyberDem0n** [comment]
> or I need to add some args into the statefulset.yam

You definitely do. Also content of such files could be stored in Secrets.

**Comment by CyberDem0n** [comment]
It was build from https://github.com/zalando/spilo/tree/feature/minispilo-10 some time ago.

**Comment by epicsoft** [comment]
Can I access the unguiculus/patroni source to update the image to egistry.opensource.zalan.do/acid/spilo-10:1.3-p9?

**Comment by CyberDem0n** [comment]
Source code of registry.opensource.zalan.do/acid/spilo-10:1.3-p9 is in https://github.com/zalando/spilo/tree/feature/minispilo-10/postgres-appliance

I've just build and pushed it 37 minutes ago.

**Comment by epicsoft** [comment]
Sorry for confusion. I refer to the source for kubernetes-native patroni: https://github.com/unguiculus/charts/tree/feature/patroni/incubator/patroni

Can I access the source of the image used in this Helm Chart?

**Comment by CyberDem0n** [comment]
> Can I access the source of the image used in this Helm Chart?

I am 100% sure that unguiculus was using https://github.com/zalando/spilo/tree/feature/minispilo-10/postgres-appliance to build and image.

**Comment by epicsoft** [comment]
Thank you very much. I think it works with your image!

**Comment by michailbrynard** [comment]
@CyberDem0n I'm not sure if this is the best place to ask, but I didn't want to start a separate issue. In both helm charts mentioned above, a patroni-config volume is mounted to etc/patroni/, however in Spilo source code, I do not see this being used anywhere. Am I missing something?

What is recommended way to update the postgres or patroni configuration once the helm chart has already been installed? E.g. if I want to update pg_max_connections in a running cluster?

**Comment by CyberDem0n** [comment]
> patroni-config volume is mounted to /etc/patroni/, however in Spilo source code, I do not see this being used anywhere. Am I missing something?

It's not used. I have no idea why it is there.

> What is recommended way to update the postgres or patroni configuration once the helm chart has already been installed? E.g. if I want to update pg_max_connections in a running cluster?

You need to exec into any running pod and use patronictl:
```bash
$ kubectl exec -ti clustername-0 su postgres
bash-4.3$ patronictl edit-config $SCOPE  # it will invoke $EDITOR, there you can change some parameters
--- 
+++ 
@@ -20,7 +20,7 @@
     log_min_duration_statement: 500
     log_statement: ddl
     log_temp_files: 0
-    max_connections: '100'
+    max_connections: '101'
     max_replication_slots: 5
     max_wal_senders: 5
     shared_preload_libraries: pg_stat_statements

Apply these changes? [y/N]: y
Configuration changed
```

After that you can use patronictl to restart pods if it is necessary:
```bash
bash-4.3$ patronictl restart $SCOPE clustername-0
```

If you are not sure whether restart is needed, you can check it:
```bash
bash-4.3$ curl -s $POD_IP:8008/patroni | jq .
{
  "state": "running",
  "postmaster_start_time": "2017-12-14 10:42:01.204 UTC",
  "role": "master",
  "server_version": 90606,
  "xlog": {
    "location": 574385630568
  },
  "replication": [
    {
      "usename": "standby",
      "application_name": "clustername-1",
      "client_addr": "10.2.71.66",
      "state": "streaming",
      "sync_state": "async",
      "sync_priority": 0
    }
  ],
  "database_system_identifier": "6388021286805409858",
  "pending_restart": true,
  "patroni": {
    "version": "1.3.6",
    "scope": "clustername"
  }
}
```

There you can see `"pending_restart": true` if restart is required.

**Comment by michailbrynard** [comment]
Thank you, that's very helpful!

**Comment by michailbrynard** [comment]
The` patronictl edit-config $SCOPE` command worked for me when I was testing with registry.opensource.zalan.do/acid/spilo-9.5, however when I run the same command in egistry.opensource.zalan.do/acid/spilo-10:1.3-p9 with Kubernetes as the DCS, I get the following error:  
```
  File "/usr/local/lib/python3.5/dist-packages/patroni/config.py", line 291, in _build_effective_configuration
    if 'authentication' in config['restapi']:
KeyError: 'restapi'
```

**Comment by michailbrynard** [comment]
I managed to figure it out, I need to pass the configuration file location as well:
> patronictl -c postgres.yml edit-config $SCOPE  
The traceback was not very helpful though.

**Comment by alexeyklyukin** [comment]
@kmoppel what if you` patronictl -c $config.yml list` for each of those nodes?

**Comment by CyberDem0n** [comment]
@kmoppel 
can you try an example from [feature/demo]( https://github.com/zalando/patroni/tree/feature/demo/kubernetes) branch?

It works without any problems on 1.10.

**Comment by kmoppel** [comment]
Thanks for the tips! But still the same with the config from demo branch - using local etcd...I'm probably doing  something wrong...but hmm what could that be? I only changed replica count from 3 to 2 in the YAML and also haven't changed minikube in any way. Do you have any idea maybe?
```
postgres@patronidemo-0:~$ patronictl -c /patroni.yml list
Error: 'Can not find suitable configuration of distributed configuration store\nAvailable implementations: zookeeper, exhibitor, etcd'
```
Just in case - env vars and processlist from a pod is [here](https://gist.github.com/kmoppel/8ea4cdb7e8225a4a25ae4ec25988dc12)

**Comment by kmoppel** [comment]
@CyberDem0n nope, just "git clone patroni" + "docker build -t patroni ."

**Comment by CyberDem0n** [comment]
@kmoppel you are building a wrong image.
There is a special Dockerfile for k8s in the `kubernetes` directory

**Comment by kmoppel** [comment]
You're right, works now from "demo" branch! Thanks a lot!
